### ISSUE TITLE
Create attribute without using @ for compatibility with json serialization

### ISSUE THREAD

**Issue by jdmichaud** [tafia/quick-xml#582]
I have a data structure with an attribute:
```rust
#[derive(Debug, Serialize, Deserialize)]
pub struct Foo {
  #[serde(rename = "@bar")]
  pub bar: String,
}
```
Which (de)serializes to:
```xml
<foo bar="blabla"/>
```
For this to work I had to add `@`. However, I need to (de)serialize in XML **and** json. So now, in json, the field appear as `@bar`:
```json
{
  "foo": {
    "@bar": "blabla",
  },
}
```
And that's a problem for me. Is there a way to mark a field as attribute without modifying its name?

**Comment by Mingun** [comment]
>  Is there a way to mark a field as attribute without modifying its name?

Unfortunately, no.

You still can solve this problem partially. You can use `#[serde(alias = "...")]` to allow deserialization both from XML and JSON (but in XML the field will be deserialized either from element or attribute). For the serialization you could write intermediate serializer, which will strip `@` from serialized fields and forward them to the underlying serializer. I think, I even accept a PR that will add such a serializer to quick-xml.

**Comment by jdmichaud** [comment]
> For the serialization you could write intermediate serializer, which will strip @ from serialized fields and forward them to the underlying serializer. I think, I even accept a PR that will add such a serializer to quick-xml.

But if I were to do that, wouldn't the underlying XML serializer serialize the field as an element?

**Comment by Mingun** [comment]
This depends on what you make as an alias, and what as a main name.
```rust
#[serde(rename = "@field")]
#[serde(alias = "field")]
```
is XML-centric (XML: serialize as attribute, deserialize from attribute or element; JSON: serialize with `@`).

```rust
#[serde(rename = "field")]
#[serde(alias = "@field")]
```
is JSON-centric (XML: serialize as element, deserialize from attribute or element; JSON: serialize without `@`).

**Comment by Mingun** [comment]
No, intermediate serializer should be used with other format serializers. The quick-xml `Serializer` should be used directly

**Comment by Nickersoft** [comment]
@Mingun I'm running up against this issue myself right now, and have been scouring the serde docs trying to find an approach I could use. I'm only interested in serialization, and modifying how data is serialized based on whether it's XML or JSON.

Could you share any examples of the intermediate serialization approach you mentioned?

**Comment by Mingun** [comment]
You can look at the `Readable` and `Compact` imlementations in `serde_test` here:
https://github.com/serde-rs/test/blob/1bedc0c7830c3207802b4b0614977c71608461cd/src/configure.rs#L181

**Comment by jbethune** [comment]
I also ran into this problem when I need both, JSON and XML. I looked a bit at the code. `serde` is driving the serialisation process and we only have limited ways of influencing it apart from `#[serde(rename=@foo)]`:

This is the trait-based callback for serde:
https://github.com/tafia/quick-xml/blob/6238d8a16179d68718b43882e544ebd20982c79c/src/se/element.rs#L516-L521

And the call chain goes here:
https://github.com/tafia/quick-xml/blob/6238d8a16179d68718b43882e544ebd20982c79c/src/se/element.rs#L386-L397

The code already has a friendly `TODO` about this problem. It would be nice if we could do something like `T::is_attribute(key)`. That would require a new trait that would also need to be added to the trait bounds for `T` (lines 388, 473, 518) and that would be a breaking change. This could also have some serious performance implications, because this code is called a lot.

This is not an easy problem to solve.

**Comment by Mingun** [comment]
This TODO is about adding a callback to the quick-xml's `Serializer` that will take `&str` and answer if it should be serialized as attribute or not.

But that is not the right way here. The right way is to create special intermediate serializer that will strip `@` from keys and pass serialization to the inser serializer. Then instead of
```rust
value.serialize(serde_json::Serializer::new(...)) // for JSON
value.serialize(quick_xml::se::Serializer::new(...)) // for XML
```
use
```rust
value.serialize(StripKeySerializer::new("@", serde_json::Serializer::new(...))) // for JSON
value.serialize(quick_xml::se::Serializer::new(...)) // for XML
```
